File: simplified-books-core/src/main/java/org/nypl/simplified/books/core/BookStatusRevoked.java
package org.nypl.simplified.books.core;

import com.io7m.jnull.NullCheck;

import java.net.URI;

/**
 * The given book is revoked, but the server has not yet been notified. A given
 * book is expected to spend very little time in this state, and will not be
 * displayed.
 */

public final class BookStatusRevoked implements BookStatusType
{
  private final BookID id;
  private final URI    revoke;

  /**
   * Construct a status value.
   *
   * @param in_id     The book ID
   * @param in_revoke The revocation URI
   */

  public BookStatusRevoked(
    final BookID in_id,
    final URI in_revoke)
  {
    this.id = NullCheck.notNull(in_id);
    this.revoke = NullCheck.notNull(in_revoke);
  }

  @Override public BookID getID()
  {
    return this.id;
  }

  @Override public BookStatusPriorityOrdering getPriority()
  {
    return BookStatusPriorityOrdering.BOOK_STATUS_LOANED;
  }

  @Override public <A, E extends Exception> A matchBookStatus(
    final BookStatusMatcherType<A, E> m)
    throws E
  {
    return m.onBookStatusRevoked(this);
  }

  @Override public String toString()
  {
    final StringBuilder b = new StringBuilder(128);
    b.append("[BookStatusRevoked ");
    b.append(this.id);
    b.append(" ");
    b.append(this.revoke);
    b.append("]");
    return NullCheck.notNull(b.toString());
  }
}

File: simplified-books-registry-api/src/main/java/org/nypl/simplified/books/book_registry/BookStatusRevoked.kt
package org.nypl.simplified.books.book_registry

import org.nypl.simplified.books.api.BookID

/**
 * The given book is revoked, but has not yet been removed from the database. A given
 * book is expected to spend very little time in this state.
 */

data class BookStatusRevoked(val id: BookID) : BookStatusType {

  override fun getID(): BookID =
    this.id

  override fun getPriority(): BookStatusPriorityOrdering =
    BookStatusPriorityOrdering.BOOK_STATUS_LOANED

  override fun <A, E : Exception> matchBookStatus(m: BookStatusMatcherType<A, E>): A =
    m.onBookStatusRevoked(this)

  override fun toString(): String {
    val b = StringBuilder(128)
    b.append("[BookStatusRevoked ")
    b.append(this.id)
    b.append("]")
    return b.toString()
  }
}
